File: backend/verifier/verify.py
"""
Standalone zero-trust verifier — implemented in Step 2.4.

MUST NOT import from app.* — intentionally duplicated crypto logic for zero-trust boundary.
"""

import argparse
import base64
import hashlib
import json
import sys
from datetime import datetime

from cryptography.hazmat.primitives.asymmetric import ed25519
from cryptography.hazmat.primitives.serialization import load_pem_public_key


def hash_text(text: str) -> str:
    """SHA-256 of UTF-8 text, lowercase hex."""
    return hashlib.sha256(text.encode("utf-8")).hexdigest()


def hash_hex_pair(left_hex: str, right_hex: str) -> str:
    """
    Parent node hash: SHA256(left_hex + right_hex) as UTF-8 concatenation.

    Both operands are lowercase hex strings — never raw bytes.
    """
    return hashlib.sha256((left_hex + right_hex).encode("utf-8")).hexdigest()


def verify_merkle_proof(
    leaf_hash: str,
    proof: list[str],
    root: str,
    leaf_index: int,
) -> bool:
    """
    Recompute root from leaf + proof using implicit L/R from leaf_index parity.

    Intentionally duplicated from app.crypto for zero-trust boundary.
    """
    current = leaf_hash
    index = leaf_index

    for sibling in proof:
        if index % 2 == 0:
            current = hash_hex_pair(current, sibling)
        else:
            current = hash_hex_pair(sibling, current)
        index //= 2

    return current == root


def canonical_json_bytes(payload: dict) -> bytes:
    """Canonical signing serialization: sorted keys, no whitespace."""
    return json.dumps(payload, sort_keys=True, separators=(",", ":")).encode("utf-8")


def load_public_key(pem: bytes) -> ed25519.Ed25519PublicKey:
    """Load Ed25519 public key from PEM bytes."""
    return load_pem_public_key(pem)


def verify_signature(
    data: bytes, signature_b64: str, public_key: ed25519.Ed25519PublicKey
) -> bool:
    """Verify base64 Ed25519 signature against bytes."""
    try:
        signature = base64.b64decode(signature_b64)
        public_key.verify(signature, data)
        return True
    except Exception:
        return False


def verify_certificate(certificate_path: str, public_key_path: str) -> tuple[bool, str]:
    """
    Verify an AnswerCertificate.

    Returns (is_valid, reason) tuple.
    """
    # Load certificate
    try:
        with open(certificate_path, "r", encoding="utf-8") as f:
            cert = json.load(f)
    except Exception as e:
        return False, f"Failed to load certificate: {e}"

    # Load public key
    try:
        with open(public_key_path, "rb") as f:
            public_key_pem = f.read()
        public_key = load_public_key(public_key_pem)
    except Exception as e:
        return False, f"Failed to load public key: {e}"

    # Extract fields
    signature = cert.get("signature")
    if not signature:
        return False, "Certificate missing signature"

    # Verify signature
    cert_copy = cert.copy()
    del cert_copy["signature"]
    try:
        payload_bytes = canonical_json_bytes(cert_copy)
    except Exception as e:
        return False, f"Failed to canonicalize certificate: {e}"

    if not verify_signature(payload_bytes, signature, public_key):
        return False, "Signature verification failed"

    # Verify each chunk
    merkle_root = cert.get("merkle_root")
    if not merkle_root:
        return False, "Certificate missing merkle_root"

    chunks = cert.get("chunks", [])
    if not chunks:
        return False, "Certificate has no chunks"

    for chunk_data in chunks:
        chunk_text = chunk_data.get("text")
        chunk_hash = chunk_data.get("hash")
        merkle_proof = chunk_data.get("merkle_proof", [])
        chunk_index = chunk_data.get("chunk_index", 0)

        if not chunk_text or not chunk_hash:
            return False, f"Chunk missing text or hash: {chunk_data}"

        # Re-hash chunk
        actual_hash = hash_text(chunk_text)
        if actual_hash != chunk_hash:
            return False, f"Chunk hash mismatch at index {chunk_index}"

        # Verify Merkle proof
        if not verify_merkle_proof(chunk_hash, merkle_proof, merkle_root, chunk_index):
            return False, f"Merkle proof verification failed for chunk {chunk_index}"

    # All checks passed
    manifest_timestamp = cert.get("manifest_timestamp", "unknown")
    return True, f"VALID — grounded in unaltered source at {manifest_timestamp}"


def main():
    parser = argparse.ArgumentParser(
        description="Verify ATTEST AnswerCertificate (zero-trust standalone verifier)"
    )
    parser.add_argument(
        "--certificate", required=True, help="Path to certificate JSON file"
    )
    parser.add_argument(
        "--public-key", required=True, help="Path to public key PEM file"
    )
    args = parser.parse_args()

    is_valid, reason = verify_certificate(args.certificate, args.public_key)

    print(reason)
    sys.exit(0 if is_valid else 1)


if __name__ == "__main__":
    main()

File: frontend/src/lib/verify.js
/**
 * Client-side certificate verification — UI-only, not backend logic.
 * Mirrors backend/verifier/verify.py for zero-trust browser verification.
 * Fetches only the public key from the server; all crypto runs locally.
 */

import * as ed from '@noble/ed25519'

async function sha256Hex(text) {
  const data = new TextEncoder().encode(text)
  const hash = await crypto.subtle.digest('SHA-256', data)
  return Array.from(new Uint8Array(hash))
    .map((byte) => byte.toString(16).padStart(2, '0'))
    .join('')
}

function hashHexPair(leftHex, rightHex) {
  const combined = leftHex + rightHex
  const data = new TextEncoder().encode(combined)
  return crypto.subtle.digest('SHA-256', data).then((hash) =>
    Array.from(new Uint8Array(hash))
      .map((byte) => byte.toString(16).padStart(2, '0'))
      .join('')
  )
}

async function verifyMerkleProof(leafHash, proof, root, leafIndex) {
  let current = leafHash
  let index = leafIndex

  for (const sibling of proof) {
    if (index % 2 === 0) {
      current = await hashHexPair(current, sibling)
    } else {
      current = await hashHexPair(sibling, current)
    }
    index = Math.floor(index / 2)
  }

  return current === root
}

function sortKeys(value) {
  if (Array.isArray(value)) {
    return value.map(sortKeys)
  }
  if (value !== null && typeof value === 'object') {
    return Object.keys(value)
      .sort()
      .reduce((sorted, key) => {
        sorted[key] = sortKeys(value[key])
        return sorted
      }, {})
  }
  return value
}

function canonicalJsonBytes(payload) {
  return new TextEncoder().encode(JSON.stringify(sortKeys(payload)))
}

function pemToRawPublicKey(pem) {
  const b64 = pem
    .replace(/-----BEGIN PUBLIC KEY-----/g, '')
    .replace(/-----END PUBLIC KEY-----/g, '')
    .replace(/\s/g, '')
  const der = Uint8Array.from(atob(b64), (char) => char.charCodeAt(0))
  return der.slice(-32)
}

function base64ToBytes(value) {
  const binary = atob(value)
  return Uint8Array.from(binary, (char) => char.charCodeAt(0))
}

async function verifySignature(payloadBytes, signatureB64, publicKeyPem) {
  try {
    const publicKey = pemToRawPublicKey(publicKeyPem)
    const signature = base64ToBytes(signatureB64)
    return await ed.verifyAsync(signature, payloadBytes, publicKey)
  } catch {
    return false
  }
}

/**
 * Verify an AnswerCertificate entirely in the browser.
 *
 * @param {object} certificate - Parsed certificate JSON
 * @param {string} publicKeyPem - PEM-encoded Ed25519 public key
 * @returns {Promise<{ok: boolean, reason: string, hash_match: boolean, proof_valid: boolean, signature_valid: boolean}>}
 */
export async function verifyCertificateClient(certificate, publicKeyPem) {
  const signature = certificate.signature
  if (!signature) {
    return {
      ok: false,
      reason: 'Certificate missing signature',
      hash_match: false,
      proof_valid: false,
      signature_valid: false,
    }
  }

  const certCopy = { ...certificate }
  delete certCopy.signature
  const payloadBytes = canonicalJsonBytes(certCopy)

  const signatureValid = await verifySignature(payloadBytes, signature, publicKeyPem)
  if (!signatureValid) {
    return {
      ok: false,
      reason: 'Signature verification failed',
      hash_match: false,
      proof_valid: false,
      signature_valid: false,
    }
  }

  const merkleRoot = certificate.merkle_root
  if (!merkleRoot) {
    return {
      ok: false,
      reason: 'Certificate missing merkle_root',
      hash_match: true,
      proof_valid: false,
      signature_valid: true,
    }
  }

  const chunks = certificate.chunks || []
  if (chunks.length === 0) {
    return {
      ok: false,
      reason: 'Certificate has no chunks',
      hash_match: true,
      proof_valid: false,
      signature_valid: true,
    }
  }

  for (const chunkData of chunks) {
    const chunkText = chunkData.text
    const chunkHash = chunkData.hash
    const merkleProof = chunkData.merkle_proof || []
    const chunkIndex = chunkData.chunk_index ?? 0

    if (!chunkText || !chunkHash) {
      return {
        ok: false,
        reason: `Chunk missing text or hash at index ${chunkIndex}`,
        hash_match: false,
        proof_valid: false,
        signature_valid: true,
      }
    }

    const actualHash = await sha256Hex(chunkText)
    if (actualHash !== chunkHash) {
      return {
        ok: false,
        reason: `Chunk hash mismatch at index ${chunkIndex}`,
        hash_match: false,
        proof_valid: false,
        signature_valid: true,
      }
    }

    const proofOk = await verifyMerkleProof(chunkHash, merkleProof, merkleRoot, chunkIndex)
    if (!proofOk) {
      return {
        ok: false,
        reason: `Merkle proof verification failed for chunk ${chunkIndex}`,
        hash_match: true,
        proof_valid: false,
        signature_valid: true,
      }
    }
  }

  const manifestTimestamp = certificate.manifest_timestamp || 'unknown'
  return {
    ok: true,
    reason: `VALID — grounded in unaltered source at ${manifestTimestamp}`,
    hash_match: true,
    proof_valid: true,
    signature_valid: true,
  }
}
